File: backend/domain/embedding/entity/space_embedding.go
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package entity

import (
	"encoding/json"
	"strconv"
)

// EmbeddingType represents the type of embedding provider
type EmbeddingType string

const (
	EmbeddingTypeOpenAI EmbeddingType = "openai"
	EmbeddingTypeArk    EmbeddingType = "ark"
	EmbeddingTypeOllama EmbeddingType = "ollama"
	EmbeddingTypeHTTP   EmbeddingType = "http"
)

// SpaceEmbedding represents the space embedding configuration entity
type SpaceEmbedding struct {
	ID            uint64        `gorm:"primaryKey;autoIncrement;comment:主键ID" json:"id"`
	SpaceID       uint64        `gorm:"not null;comment:空间ID" json:"space_id"`
	UserID        uint64        `gorm:"not null;comment:创建者ID" json:"user_id"`
	Name          string        `gorm:"size:128;not null;comment:配置名称" json:"name"`
	Description   string        `gorm:"size:512;default:'';comment:配置描述" json:"description"`
	EmbeddingType EmbeddingType `gorm:"size:32;not null;comment:Embedding类型" json:"embedding_type"`
	Config        string        `gorm:"type:json;not null;comment:Embedding配置" json:"config"`
	MaxBatchSize  int           `gorm:"not null;default:100;comment:最大批处理大小" json:"max_batch_size"`
	Status        int           `gorm:"not null;default:1;comment:状态: 1启用 2禁用" json:"status"`
	IsDefault     int           `gorm:"not null;default:0;comment:是否为默认配置" json:"is_default"`
	CreatedAt     uint64        `gorm:"not null;default:0;comment:创建时间" json:"created_at"`
	UpdatedAt     uint64        `gorm:"not null;default:0;comment:更新时间" json:"updated_at"`
	DeletedAt     *uint64       `gorm:"comment:删除时间" json:"deleted_at"`
}

func (SpaceEmbedding) TableName() string {
	return "space_embedding"
}

// Status constants
const (
	SpaceEmbeddingStatusEnabled  = 1
	SpaceEmbeddingStatusDisabled = 2
)

// OpenAIEmbeddingConfig represents OpenAI embedding configuration
type OpenAIEmbeddingConfig struct {
	BaseURL     string `json:"base_url"`
	APIKey      string `json:"api_key"`
	Model       string `json:"model"`
	ByAzure     bool   `json:"by_azure,omitempty"`
	APIVersion  string `json:"api_version,omitempty"`
	Dims        int    `json:"dims"`
	RequestDims int    `json:"request_dims,omitempty"`
}

// ArkEmbeddingConfig represents ARK embedding configuration
type ArkEmbeddingConfig struct {
	BaseURL string `json:"base_url,omitempty"`
	APIKey  string `json:"api_key"`
	Model   string `json:"model"`
	Dims    int    `json:"dims"`
	APIType string `json:"api_type,omitempty"` // text or multimodal
}

// OllamaEmbeddingConfig represents Ollama embedding configuration
type OllamaEmbeddingConfig struct {
	BaseURL string `json:"base_url"`
	Model   string `json:"model"`
	Dims    int    `json:"dims"`
}

// HTTPEmbeddingConfig represents HTTP embedding configuration
type HTTPEmbeddingConfig struct {
	Addr string `json:"addr"`
	Dims int    `json:"dims"`
}

// EmbeddingConfig represents the unified embedding configuration
type EmbeddingConfig struct {
	Type         EmbeddingType          `json:"type"`
	MaxBatchSize int                    `json:"max_batch_size,omitempty"`
	OpenAI       *OpenAIEmbeddingConfig `json:"openai_config,omitempty"`
	Ark          *ArkEmbeddingConfig    `json:"ark_config,omitempty"`
	Ollama       *OllamaEmbeddingConfig `json:"ollama_config,omitempty"`
	HTTP         *HTTPEmbeddingConfig   `json:"http_config,omitempty"`
}

// GetConfigStruct parses the JSON config string into EmbeddingConfig
func (s *SpaceEmbedding) GetConfigStruct() (*EmbeddingConfig, error) {
	var config EmbeddingConfig
	if err := json.Unmarshal([]byte(s.Config), &config); err != nil {
		return nil, err
	}
	config.Type = s.EmbeddingType
	config.MaxBatchSize = s.MaxBatchSize
	return &config, nil
}

// SetConfigFromStruct converts EmbeddingConfig to JSON string and sets it
func (s *SpaceEmbedding) SetConfigFromStruct(config *EmbeddingConfig) error {
	data, err := json.Marshal(config)
	if err != nil {
		return err
	}
	s.Config = string(data)
	s.EmbeddingType = config.Type
	if config.MaxBatchSize > 0 {
		s.MaxBatchSize = config.MaxBatchSize
	}
	return nil
}

// GetDimensions returns the embedding dimensions based on config type
func (s *SpaceEmbedding) GetDimensions() int {
	config, err := s.GetConfigStruct()
	if err != nil {
		return 0
	}

	switch s.EmbeddingType {
	case EmbeddingTypeOpenAI:
		if config.OpenAI != nil {
			return config.OpenAI.Dims
		}
	case EmbeddingTypeArk:
		if config.Ark != nil {
			return config.Ark.Dims
		}
	case EmbeddingTypeOllama:
		if config.Ollama != nil {
			return config.Ollama.Dims
		}
	case EmbeddingTypeHTTP:
		if config.HTTP != nil {
			return config.HTTP.Dims
		}
	}
	return 0
}

// SpaceEmbeddingView is a view model for API responses
type SpaceEmbeddingView struct {
	ID            string           `json:"id"`
	SpaceID       string           `json:"space_id"`
	Name          string           `json:"name"`
	Description   string           `json:"description"`
	EmbeddingType EmbeddingType    `json:"embedding_type"`
	Config        *EmbeddingConfig `json:"config"`
	Status        int              `json:"status"`
	IsDefault     bool             `json:"is_default"`
	Dimensions    int              `json:"dimensions"`
	CreatedAt     uint64           `json:"created_at"`
	UpdatedAt     uint64           `json:"updated_at"`
}

// ToView converts SpaceEmbedding to SpaceEmbeddingView
func (s *SpaceEmbedding) ToView() *SpaceEmbeddingView {
	config, _ := s.GetConfigStruct()

	// Mask sensitive information
	if config != nil {
		if config.OpenAI != nil && config.OpenAI.APIKey != "" {
			config.OpenAI.APIKey = maskAPIKey(config.OpenAI.APIKey)
		}
		if config.Ark != nil && config.Ark.APIKey != "" {
			config.Ark.APIKey = maskAPIKey(config.Ark.APIKey)
		}
	}

	return &SpaceEmbeddingView{
		ID:            formatID(s.ID),
		SpaceID:       formatID(s.SpaceID),
		Name:          s.Name,
		Description:   s.Description,
		EmbeddingType: s.EmbeddingType,
		Config:        config,
		Status:        s.Status,
		IsDefault:     s.IsDefault == 1,
		Dimensions:    s.GetDimensions(),
		CreatedAt:     s.CreatedAt,
		UpdatedAt:     s.UpdatedAt,
	}
}

// maskAPIKey masks the API key for security
func maskAPIKey(key string) string {
	if len(key) <= 8 {
		return "****"
	}
	return key[:4] + "****" + key[len(key)-4:]
}

// formatID converts uint64 to string
func formatID(id uint64) string {
	return strconv.FormatUint(id, 10)
}

File: frontend/packages/arch/api-schema/src/idl/embedding/space_embedding.ts
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { createAPI } from './../../api/config';

/** Embedding类型 */
export enum EmbeddingType {
  OpenAI = 'openai',
  Ark = 'ark',
  Ollama = 'ollama',
  HTTP = 'http',
}

/** Embedding状态 */
export enum EmbeddingStatus {
  Enabled = 1,
  Disabled = 2,
}

/** OpenAI Embedding配置 */
export interface OpenAIEmbeddingConfig {
  base_url: string;
  api_key: string;
  model: string;
  by_azure?: boolean;
  api_version?: string;
  dims: number;
  request_dims?: number;
}

/** ARK Embedding配置 */
export interface ArkEmbeddingConfig {
  base_url?: string;
  api_key: string;
  model: string;
  dims: number;
  api_type?: string;
}

/** Ollama Embedding配置 */
export interface OllamaEmbeddingConfig {
  base_url: string;
  model: string;
  dims: number;
}

/** HTTP Embedding配置 */
export interface HttpEmbeddingConfig {
  addr: string;
  dims: number;
}

/** 统一Embedding配置 */
export interface EmbeddingConfig {
  type: EmbeddingType;
  max_batch_size?: number;
  openai_config?: OpenAIEmbeddingConfig;
  ark_config?: ArkEmbeddingConfig;
  ollama_config?: OllamaEmbeddingConfig;
  http_config?: HttpEmbeddingConfig;
}

/** 空间Embedding配置 */
export interface SpaceEmbeddingConfig {
  id: string;
  space_id: string;
  name: string;
  description?: string;
  embedding_type: EmbeddingType;
  config?: EmbeddingConfig;
  status: EmbeddingStatus;
  is_default: boolean;
  dimensions: number;
  created_at: number;
  updated_at: number;
}

/** 创建空间Embedding配置请求 */
export interface CreateSpaceEmbeddingRequest {
  space_id: string;
  name: string;
  description?: string;
  type: EmbeddingType;
  config: Record<string, any>;
  max_batch_size?: number;
  set_as_default?: boolean;
}

export interface CreateSpaceEmbeddingResponse {
  code: number;
  msg: string;
  data?: SpaceEmbeddingConfig;
}

/** 获取空间Embedding列表请求 */
export interface ListSpaceEmbeddingsRequest {
  space_id: string;
}

export interface ListSpaceEmbeddingsResponse {
  code: number;
  msg: string;
  data?: SpaceEmbeddingConfig[];
}

/** 获取空间默认Embedding请求 */
export interface GetSpaceDefaultEmbeddingRequest {
  space_id: string;
}

export interface GetSpaceDefaultEmbeddingResponse {
  code: number;
  msg: string;
  data?: SpaceEmbeddingConfig;
}

/** 更新空间Embedding配置请求 */
export interface UpdateSpaceEmbeddingRequest {
  space_id: string;
  embedding_id: string;
  name?: string;
  description?: string;
  type?: EmbeddingType;
  config?: Record<string, any>;
  max_batch_size?: number;
}

export interface UpdateSpaceEmbeddingResponse {
  code: number;
  msg: string;
  data?: SpaceEmbeddingConfig;
}

/** 删除空间Embedding配置请求 */
export interface DeleteSpaceEmbeddingRequest {
  space_id: string;
  embedding_id: string;
}

export interface DeleteSpaceEmbeddingResponse {
  code: number;
  msg: string;
}

/** 设置默认Embedding请求 */
export interface SetDefaultSpaceEmbeddingRequest {
  space_id: string;
  embedding_id: string;
}

export interface SetDefaultSpaceEmbeddingResponse {
  code: number;
  msg: string;
}

/** 启用/禁用Embedding请求 */
export interface EnableSpaceEmbeddingRequest {
  space_id: string;
  embedding_id: string;
}

export interface EnableSpaceEmbeddingResponse {
  code: number;
  msg: string;
}

export interface DisableSpaceEmbeddingRequest {
  space_id: string;
  embedding_id: string;
}

export interface DisableSpaceEmbeddingResponse {
  code: number;
  msg: string;
}

// API 函数定义

export const CreateSpaceEmbedding = /*#__PURE__*/createAPI<CreateSpaceEmbeddingRequest, CreateSpaceEmbeddingResponse>({
  url: '/api/embedding/space/create',
  method: 'POST',
  name: 'CreateSpaceEmbedding',
  reqType: 'CreateSpaceEmbeddingRequest',
  reqMapping: {
    body: ['space_id', 'name', 'description', 'type', 'config', 'max_batch_size', 'set_as_default'],
  },
  resType: 'CreateSpaceEmbeddingResponse',
  schemaRoot: 'api://schemas/idl_embedding_space_embedding',
  service: 'space_embedding',
});

export const ListSpaceEmbeddings = /*#__PURE__*/createAPI<ListSpaceEmbeddingsRequest, ListSpaceEmbeddingsResponse>({
  url: '/api/embedding/space/list',
  method: 'POST',
  name: 'ListSpaceEmbeddings',
  reqType: 'ListSpaceEmbeddingsRequest',
  reqMapping: {
    body: ['space_id'],
  },
  resType: 'ListSpaceEmbeddingsResponse',
  schemaRoot: 'api://schemas/idl_embedding_space_embedding',
  service: 'space_embedding',
});

export const GetSpaceDefaultEmbedding = /*#__PURE__*/createAPI<GetSpaceDefaultEmbeddingRequest, GetSpaceDefaultEmbeddingResponse>({
  url: '/api/embedding/space/default',
  method: 'POST',
  name: 'GetSpaceDefaultEmbedding',
  reqType: 'GetSpaceDefaultEmbeddingRequest',
  reqMapping: {
    body: ['space_id'],
  },
  resType: 'GetSpaceDefaultEmbeddingResponse',
  schemaRoot: 'api://schemas/idl_embedding_space_embedding',
  service: 'space_embedding',
});

export const UpdateSpaceEmbedding = /*#__PURE__*/createAPI<UpdateSpaceEmbeddingRequest, UpdateSpaceEmbeddingResponse>({
  url: '/api/embedding/space/update',
  method: 'POST',
  name: 'UpdateSpaceEmbedding',
  reqType: 'UpdateSpaceEmbeddingRequest',
  reqMapping: {
    body: ['space_id', 'embedding_id', 'name', 'description', 'type', 'config', 'max_batch_size'],
  },
  resType: 'UpdateSpaceEmbeddingResponse',
  schemaRoot: 'api://schemas/idl_embedding_space_embedding',
  service: 'space_embedding',
});

export const DeleteSpaceEmbedding = /*#__PURE__*/createAPI<DeleteSpaceEmbeddingRequest, DeleteSpaceEmbeddingResponse>({
  url: '/api/embedding/space/delete',
  method: 'POST',
  name: 'DeleteSpaceEmbedding',
  reqType: 'DeleteSpaceEmbeddingRequest',
  reqMapping: {
    body: ['space_id', 'embedding_id'],
  },
  resType: 'DeleteSpaceEmbeddingResponse',
  schemaRoot: 'api://schemas/idl_embedding_space_embedding',
  service: 'space_embedding',
});

export const SetDefaultSpaceEmbedding = /*#__PURE__*/createAPI<SetDefaultSpaceEmbeddingRequest, SetDefaultSpaceEmbeddingResponse>({
  url: '/api/embedding/space/set-default',
  method: 'POST',
  name: 'SetDefaultSpaceEmbedding',
  reqType: 'SetDefaultSpaceEmbeddingRequest',
  reqMapping: {
    body: ['space_id', 'embedding_id'],
  },
  resType: 'SetDefaultSpaceEmbeddingResponse',
  schemaRoot: 'api://schemas/idl_embedding_space_embedding',
  service: 'space_embedding',
});

export const EnableSpaceEmbedding = /*#__PURE__*/createAPI<EnableSpaceEmbeddingRequest, EnableSpaceEmbeddingResponse>({
  url: '/api/embedding/space/enable',
  method: 'POST',
  name: 'EnableSpaceEmbedding',
  reqType: 'EnableSpaceEmbeddingRequest',
  reqMapping: {
    body: ['space_id', 'embedding_id'],
  },
  resType: 'EnableSpaceEmbeddingResponse',
  schemaRoot: 'api://schemas/idl_embedding_space_embedding',
  service: 'space_embedding',
});

export const DisableSpaceEmbedding = /*#__PURE__*/createAPI<DisableSpaceEmbeddingRequest, DisableSpaceEmbeddingResponse>({
  url: '/api/embedding/space/disable',
  method: 'POST',
  name: 'DisableSpaceEmbedding',
  reqType: 'DisableSpaceEmbeddingRequest',
  reqMapping: {
    body: ['space_id', 'embedding_id'],
  },
  resType: 'DisableSpaceEmbeddingResponse',
  schemaRoot: 'api://schemas/idl_embedding_space_embedding',
  service: 'space_embedding',
});
